runtime/src/main/java/org/corfudb/runtime/object/transactions/OptimisticTransactionalContext.java

```diff
@@ -262,41 +272,190 @@ public long getConflictSetAndCommit(ConflictSetInfo conflictSet) {
 
         super.commitTransaction();
         commitAddress = address;
 
         tryCommitAllProxies();
-        log.trace("Commit[{}] Written to {}", this, address);
+        log.trace("doCommit[{}] Written to {}", this, address);
         return address;
     }
 
+    /** Do a precise commit, which is guaranteed not to produce an abort
+     * due to a false conflict. This method achieves this guarantee by scanning
+     * scanning the log when the sequencer detects a conflict, and manually
+     * inspecting each entry in the conflict window. If there are no true conflicts,
+     * the transaction is retried with the sequencer, otherwise, if a true conflict
+     * is detected, the transaction aborts, with a flag indicating the the abort
+     * was precise (guaranteed to not be false).
+     *
+     * @param originalException     The original exception when we first tried to commit.
+     * @param conflictSet           The set of objects this transaction conflicts with.
+     * @param hashedConflictSet     The hashed version of the conflict set.
+     * @param affectedStreams       The set of streams affected by this transaction.
+     * @return                      The address the transaction was committed to.
+     * @throws TransactionAbortedException  If the transaction must be aborted.
+     */
+    protected long doPreciseCommit(@Nonnull final TransactionAbortedException originalException,
+                                   @NonNull final Map<ICorfuSMRProxyInternal, Set<Object>>
+                                           conflictSet,
+                                   @NonNull final Map<UUID, Set<byte[]>> hashedConflictSet,
+                                   @NonNull final Set<UUID> affectedStreams) {
+
+        log.debug("doPreciseCommit[{}]: Imprecise conflict detected, resolving...", this);
+        TransactionAbortedException currentException = originalException;
+
+        // This map contains the maximum address of the streams we have
+        // verified to not have any true conflicts so far.
+        final Map<UUID, Long> verifiedStreams = new HashMap<>();
+
+        // We resolve conflicts until we have a -true- conflict.
+        // This might involve having the sequencer reject our
+        // request to commit multiple times.
+        while (currentException.getAbortCause() == AbortCause.CONFLICT) {
+            final UUID conflictStream = currentException.getConflictStream();
+            final long currentAddress = currentException.getConflictAddress();
+            final TransactionAbortedException thisException = currentException;
+
+            // Get the proxy, which should be available either in the write set
+            // or read set. We need the proxy to generate the conflict objects
+            // from the SMR entry.
+            ICorfuSMRProxyInternal proxy;
+            Optional<ICorfuSMRProxyInternal> modifyProxy = getModifiedProxies().stream()
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "ICorfuSMRProxyInternal"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)